fix: bulkWriter: writing to the same document does not create a new batch

thebrianchen opened the pull request.

Commit:
fix: bulkWriter: writing to the same document does not create a new batch

Commit:
remove now-redundant system test

schmidt-sebastian commented on dev/src/bulk-writer.ts:
```diff
@@ -229,56 +235,62 @@ class BulkCommitBatch {
-      this.pendingOps.get(result.key)!.reject(result.status);
-      this.pendingOps.delete(result.key);
+    for (let i = resolvedOps.length - 1; i >= 0; i--) {
+      this.pendingOps.splice(resolvedOps[i], 1);
```
Aren't these indices off after the first splice? I wonder if we can simply build up the reverse array in the loop above and replace this.pendingOps with it.

thebrianchen replied:
Since we're going in reverse order and deleting the last element in the array each time, the remaining indices won't be affected.

schmidt-sebastian replied:
Can you add a comment?

thebrianchen replied:
done.

schmidt-sebastian commented on dev/test/bulk-writer.ts:
```diff
@@ -348,23 +332,20 @@ describe('BulkWriter', () => {
   });
 
-  it('sends writes to the same document in separate batches', async () => {
+  it('sends writes to the same documents in the same batch', async () => {
```
Does this test still have value over the other tests?

thebrianchen replied:
This test would fail if pendingOps got changed back to its original map type since the second write to the same document path would override the existing entry, so it serves to make sure that writes to the same document are added properly.

schmidt-sebastian replied:
Makes sense. Maybe prefix the test name with "can send"?

thebrianchen replied:
done.

schmidt-sebastian reviewed (commented):
This is mostly good, though I wonder if we need to fix the "docsToRetry" logic in the WriteBatch constructor?

Commit:
add comments

schmidt-sebastian reviewed (commented):
Thanks for the updates. Can you find a way to address the comment in the summary response?

Commit:
Update dev/src/bulk-writer.ts

Co-authored-by: Sebastian Schmidt <mrschmidt@google.com>

Commit:
replace docsToRetry with indexesToRetry

thebrianchen commented:
> Thanks for the updates. Can you find a way to address the comment in the summary response?

Updated to use an index-based tracking system instead. Also modified the retry test to include retries to the same document path, which would have failed with the previous iteration.

schmidt-sebastian commented on dev/src/bulk-writer.ts:
```diff
@@ -229,75 +241,76 @@ class BulkCommitBatch {
-        this.pendingOps.delete(result.key);
+  private processResults(
+    results: BatchWriteResult[],
+    failRemainingOperations = false
```
I think this should be "allowRetry" or something like that, which also means that you can just merge the corresponding branch with `if (!this.shouldRetry(result.status.code))`.  In the context of this method, `failRemainingOperations` is quite self-explanatory.

thebrianchen replied:
Done.

schmidt-sebastian commented on dev/src/bulk-writer.ts:
```diff
@@ -229,75 +241,76 @@ class BulkCommitBatch {
-        this.pendingOps.delete(result.key);
+        op.deferred.reject(result.status);
+      } else {
+        // Retry the operation if it is not processed.
```
Maybe: 

```suggestion
        // Retry the operation if it has not been processed.
```

schmidt-sebastian commented on dev/src/write-batch.ts:
```diff
@@ -144,29 +144,29 @@ export class WriteBatch implements firestore.WriteBatch {
+      // indexes to retry.
       this._ops = retryBatch._ops.filter(
-        v => docsToRetry!.indexOf(v.docPath) !== -1
+        (op, index) => indexesToRetry!.indexOf(index) !== -1
```
This is a bit confusing to read. Could we make indexesToRetry a `Set`? Optional.

thebrianchen replied:
Good call. Done.

Commit:
Merge branch 'master' into bc/bulk-no-order

Commit:
Update dev/src/bulk-writer.ts

Co-authored-by: Sebastian Schmidt <mrschmidt@google.com>

Commit:
use allowRetry and use a Set for indexesToRetry
